src/utils/daos_hdlr.c

```diff
@@ -1645,15 +1651,93 @@ fs_copy_file(struct cmd_args_s *ap,
 	return rc;
 }
 
+static int
+fs_copy_symlink(struct cmd_args_s *ap,
+		struct file_dfs *src_file_dfs,
+		struct file_dfs *dst_file_dfs,
+		struct stat *src_stat,
+		const char *src_path,
+		const char *dst_path)
+{
+	int		rc = 0;
+	daos_size_t	len = DFS_MAX_PATH;
+	char		*symlink_value;
+
+	D_ALLOC(symlink_value, len + 1);
+	if (symlink_value == NULL)
+		D_GOTO(out_copy_symlink, rc = -DER_NOMEM);
+
+	if (src_file_dfs->type == POSIX) {
+		len = readlink(src_path, symlink_value, (size_t)len);
+		if (len < 0) {
+			rc = daos_errno2der(errno);
+			DH_PERROR_DER(ap, rc, "fs_copy_symlink failed on readlink('%s')",
+				      src_path);
+			D_GOTO(out_copy_symlink, rc);
+		}
+	} else if (src_file_dfs->type == DAOS) {
+		rc = dfs_sys_readlink(src_file_dfs->dfs_sys, src_path, symlink_value, &len);
+		if (rc != 0) {
+			rc = daos_errno2der(rc);
+			DH_PERROR_DER(ap, rc, "fs_copy_symlink failed on dfs_sys_readlink('%s')",
+				      src_path);
+			D_GOTO(out_copy_symlink, rc);
+		}
+		/*length of symlink_value includes the NULL terminator already.*/
+		len--;
+	} else {
+		rc = -DER_INVAL;
+		DH_PERROR_DER(ap, rc, "unknown type for %s", src_path);
+		D_GOTO(out_copy_symlink, rc);
+	}
+	symlink_value[len] = 0;
+
+	if (dst_file_dfs->type == POSIX) {
+		rc = symlink(symlink_value, dst_path);
+		if ((rc != 0) && (errno == EEXIST)) {
+			rc = -DER_EXIST;
+			D_DEBUG(DB_TRACE, "Link %s exists.\n", dst_path);
+			D_GOTO(out_copy_symlink, rc);
+		} else if (rc != 0) {
+			rc = daos_errno2der(errno);
+			DH_PERROR_DER(ap, rc, "fs_copy_symlink failed on symlink('%s')",
+				      dst_path);
+			D_GOTO(out_copy_symlink, rc);
+		}
+	} else if (dst_file_dfs->type == DAOS) {
+		rc = dfs_sys_symlink(dst_file_dfs->dfs_sys, symlink_value, dst_path);
+		if (rc == EEXIST) {
+			rc = -DER_EXIST;
+			D_DEBUG(DB_TRACE, "Link %s exists.\n", dst_path);
+			D_GOTO(out_copy_symlink, rc);
+		} else if (rc != 0) {
+			rc = daos_errno2der(rc);
+			DH_PERROR_DER(ap, rc,
+				      "fs_copy_symlink failed on dfs_sys_readlink('%s')",
+				      dst_path);
+			D_GOTO(out_copy_symlink, rc);
+		}
+	} else {
+		rc = -DER_INVAL;
+		DH_PERROR_DER(ap, rc, "unknown type for %s", dst_path);
+		D_GOTO(out_copy_symlink, rc);
+	}
+out_copy_symlink:
+	if (symlink_value)
+		D_FREE(symlink_value);
```

Comment:
Minor, but we don't normally check for NULL before freeing things.
```suggestion
	D_FREE(symlink_value);
```